bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/ItemChannelLinkRegistry.java

```diff
@@ -99,19 +198,40 @@ public void removeLinksForThing(ThingUID thingUID) {
 
     @Override
     protected void notifyListenersAboutAddedElement(ItemChannelLink element) {
-        super.notifyListenersAboutAddedElement(element);
-        postEvent(LinkEventFactory.createItemChannelLinkAddedEvent(element));
+        notifyListenersAboutAddedElement(element, itemRegistry.containsItem(element.getItemName()));
     }
 
     @Override
     protected void notifyListenersAboutRemovedElement(ItemChannelLink element) {
-        super.notifyListenersAboutRemovedElement(element);
-        postEvent(LinkEventFactory.createItemChannelLinkRemovedEvent(element));
+        // Workaround for removeProvider: see description of member 'active' above
+        final ItemRegistry itemRegistry = this.itemRegistry;
+        final boolean contains;
+        if (itemRegistry != null) {
+            contains = itemRegistry.containsItem(element.getItemName());
+        } else {
+            contains = true;
+        }
+        notifyListenersAboutRemovedElement(element, contains);
     }
 
     @Override
     protected void notifyListenersAboutUpdatedElement(ItemChannelLink oldElement, ItemChannelLink element) {
         super.notifyListenersAboutUpdatedElement(oldElement, element);
         // it is not needed to send an event, because links can not be updated
     }
+
+    protected void notifyListenersAboutAddedElement(final ItemChannelLink element, boolean itemExists) {
+        if (itemExists) {
+            super.notifyListenersAboutAddedElement(element);
```

Comment:
You now completely ignore the fact that links are created by even suppressing their events. I don't think we should do that. The link is there when it is created, it should merely not be assumed that the item exists as well.

I would expect that the ChannelItemProvider stops working with your current code, since it actually listens for these events and only creates its items when a new link is created.